### PR TITLE
C++: IR dataflow through modeled functions

### PR DESCRIPTION
**Pull request by rdmarsh2**

### CODE REVIEW

**Comment by jbj** [comment]
Does `DefaultTaintTracking` benefit from this?

**Comment by jbj** [comment]
These changes that you're copying from `DefaultTaintTracking` to `DataFlow` are full of conflation between pointer and object. Wouldn't it be best to keep that conflation in the taint-tracking libraries?

**Comment by rdmarsh2** [comment]
> Does `DefaultTaintTracking` benefit from this?

It should, since it doesn't use `DataFlowFunction`, only `TaintFunction`. I'm not sure if any of the existing tests will be affected by it (they weren't when I first submitted it, but newer PRs might have changed that).

> These changes that you're copying from `DefaultTaintTracking` to `DataFlow` are full of conflation between pointer and object. Wouldn't it be best to keep that conflation in the taint-tracking libraries?

Yes. I'll rewrite to remove that conflation where we have side effect instructions for both (return values and qualifiers don't currently have side effect instructions.)

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll [hunk 1]
```diff
@@ -172,7 +172,7 @@ private predicate instructionTaintStep(Instruction i1, Instruction i2) {
     any(CallInstruction call |
       exists(int indexIn |
         modelTaintToReturnValue(call.getStaticCallTarget(), indexIn) and
-        i1 = getACallArgumentOrIndirection(call, indexIn)
+        i1 = DataFlow::getACallArgumentOrIndirection(call, indexIn)
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll]: It looks like there's no longer a need to copy this predicate to `DataFlowUtil.qll` after your latest changes.

**Pull request merged**